### ISSUE TITLE
Can't set headers after they are sent.

### ISSUE THREAD

**Issue by MuslimMunir** [balderdashy/sails#4222]
I have observed that this error is also thrown when the sails app is halted for a while and the response is not being sent.
Like the case I have tried is I have given wrong mysql server credentials or the credentials of server to which i don't have access allowed.
What happened next is this error

Error occurred in session middleware :: 'Error: Handshake inactivity timeout\n at Handshake. (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:160:17)\n at emitNone (events.js:86:13)\n at Handshake.emit (events.js:185:7)\n at Handshake._onTimeout (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/sequences/Sequence.js:126:8)\n at tryOnTimeout (timers.js:232:11)\n at Timer.listOnTimeout (timers.js:202:5)\n --------------------\n at Protocol._enqueue (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:141:48)\n at Protocol.handshake (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:52:41)\n at PoolConnection.connect (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Connection.js:136:18)\n at Pool.getConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:48:16)\n at Pool.releaseConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:155:10)\n at Pool._removeConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:275:8)\n at Pool._purgeConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:256:8)\n at Handshake.onConnect (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:57:14)\n at Handshake.Sequence.end (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/sequences/Sequence.js:85:24)\n at /home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:399:18\n at Array.forEach (native)\n at /home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:398:13\n at _combinedTickCallback (internal/process/next_tick.js:67:7)\n at process._tickDomainCallback (internal/process/next_tick.js:122:9)'

and after that I tried to hit that same API with loadtest thousands of time and program crashed with below error.

_http_outgoing.js:344
    throw new Error('Can\'t set headers after they are sent.');
    ^
Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:344:11)
    at ServerResponse.res.setHeader (/usr/lib/node_modules/sails/node_modules/@sailshq/connect/lib/patch.js:134:22)
    at ServerResponse.res.set.res.header (/usr/lib/node_modules/sails/node_modules/@sailshq/express/lib/response.js:595:10)
    at ServerResponse.res.send (/usr/lib/node_modules/sails/node_modules/@sailshq/express/lib/response.js:151:12)

**Sails version**:0.12.4
**Node version**:6.11.4
**NPM version**:3.10.10
**DB adapter name**:sails-mysql
**DB adapter version**: 0.11.5
**Operating system**:14.04 ubuntu

**Comment by sailsbot** [comment]
Hi @MuslimMunir!  It looks like you may have removed some required elements from the initial comment template, without which I can't verify that this post meets our [contribution guidelines](http://bit.ly/sails-issue-guide). To re-open this issue, please copy the template from [here](https://raw.githubusercontent.com/balderdashy/sails/master/.github/ISSUE_TEMPLATE), paste it at the beginning of your initial comment, and follow the instructions in the text. Then post a new comment (e.g. "ok, fixed!") so that I know to go back and check.

Sorry to be a hassle, but following these instructions ensures that we can help you in the best way possible and keep the Sails project running smoothly.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by MuslimMunir** [comment]
### BEGIN PLEDGE ###
- [X] I am experiencing a concrete technical issue (aka a bug) with Sails (ideas and feature proposals should follow the guide for proposing features and enhancements (http://bit.ly/sails-feature-guide), which involves making a pull request).  If you're not 100% certain whether it's a bug or not, that's okay--you may continue.  The worst that can happen is that the issue will be closed and we'll point you in the right direction.

- [X] I am not asking a question about how to use Sails or about whether or not Sails has a certain feature (please refer to the documentation(http://sailsjs.com), or post on http://stackoverflow.com, our Google Group (http://bit.ly/sails-google-group) or our live chat (https://gitter.im/balderdashy/sails).

- [X] I have already searched for related issues, and found none open (if you found a related _closed_ issue, please link to it in your post).

- [X] My issue title is concise, on-topic and polite ("jst.js being removed from layout.ejs on lift" is good; "templates dont work" or "why is sails dumb" are not so good).

- [X] I have tried all the following (if relevant) and my issue remains:
  - Make sure you have the right app lifted.
  - Make sure you've killed the Sails server with CTRL+C and started it again.
  - Make sure you closed any open browser tabs pointed at localhost before starting Sails.
  - Make sure you do not have any other Sails apps running in other terminal windows.
  - Make sure the app you are using to reproduce the issue has a clean node_modules/ directory, meaning:
    * no dependencies are linked (e.g. you haven't run npm link foo)
    * that you haven't made any inline changes to files in the node_modules/ folder
    * that you don't have any weird global dependency loops. The easiest way to double-check any of the above, if you aren't sure, is to run: rm -rf node_modules && npm cache clear && npm install.

and finally...
- [X] **I can provide steps to reproduce this issue that others can follow.**

Ideally, this involves creating a new repo that demonstrates the problem (see instructions at http://bit.ly/sails-issue-repro).  Even though your issue may seem so simple to reproduce that a new repo is unnecessary, you'd be surprised how many solutions present themselves when you start from `sails new` and attempt to recreate your issue from scratch in a new app.  This ensures that the real issue isn't in your user code (a forgotten policy file, perhaps?) or in a third-party module.  If you're absolutely _convinced_ that a new repo is unnecessary, provide clear, concise and _specific_ steps to reproduce the problem in your post (not "create a model then do blueprint create").
### END PLEDGE ###

I have observed that this error is also thrown when the sails app is halted for a while and the response is not being sent.
Like the case I have tried is I have given wrong mysql server credentials or the credentials of server to which i don't have access allowed.
What happened next is this error

Error occurred in session middleware :: 'Error: Handshake inactivity timeout\n at Handshake. (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:160:17)\n at emitNone (events.js:86:13)\n at Handshake.emit (events.js:185:7)\n at Handshake._onTimeout (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/sequences/Sequence.js:126:8)\n at tryOnTimeout (timers.js:232:11)\n at Timer.listOnTimeout (timers.js:202:5)\n --------------------\n at Protocol._enqueue (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:141:48)\n at Protocol.handshake (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:52:41)\n at PoolConnection.connect (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Connection.js:136:18)\n at Pool.getConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:48:16)\n at Pool.releaseConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:155:10)\n at Pool._removeConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:275:8)\n at Pool._purgeConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:256:8)\n at Handshake.onConnect (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:57:14)\n at Handshake.Sequence.end (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/sequences/Sequence.js:85:24)\n at /home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:399:18\n at Array.forEach (native)\n at /home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:398:13\n at _combinedTickCallback (internal/process/next_tick.js:67:7)\n at process._tickDomainCallback (internal/process/next_tick.js:122:9)'

and after that I tried to hit that same API with loadtest thousands of time and program crashed with below error.

_http_outgoing.js:344
    throw new Error('Can\'t set headers after they are sent.');
    ^
Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:344:11)
    at ServerResponse.res.setHeader (/usr/lib/node_modules/sails/node_modules/@sailshq/connect/lib/patch.js:134:22)
    at ServerResponse.res.set.res.header (/usr/lib/node_modules/sails/node_modules/@sailshq/express/lib/response.js:595:10)
    at ServerResponse.res.send (/usr/lib/node_modules/sails/node_modules/@sailshq/express/lib/response.js:151:12)


**Sails version**:0.12.4
**Node version**:6.11.4
**NPM version**:3.10.10
**DB adapter name**:sails-mysql
**DB adapter version**: 0.11.5
**Operating system**:14.04 ubuntu

**Comment by MuslimMunir** [comment]
**Sails version**:0.12.4
**Node version**:6.11.4
**NPM version**:3.10.10
**DB adapter name**: sails-mysql
**DB adapter version**: 0.11.5
**Operating system**:14.04 ubuntu


<hr/>

**Comment by sailsbot** [comment]
Hi @MuslimMunir!  It looks like you may have removed some required elements from the initial comment template, without which I can't verify that this post meets our [contribution guidelines](http://bit.ly/sails-issue-guide). To re-open this issue, please copy the template from [here](https://raw.githubusercontent.com/balderdashy/sails/master/.github/ISSUE_TEMPLATE), paste it at the beginning of your initial comment, and follow the instructions in the text. Then post a new comment (e.g. "ok, fixed!") so that I know to go back and check.

Sorry to be a hassle, but following these instructions ensures that we can help you in the best way possible and keep the Sails project running smoothly.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by sailsbot** [comment]
Hi @MuslimMunir!  It looks like you may have removed some required elements from the initial comment template, without which I can't verify that this post meets our [contribution guidelines](http://bit.ly/sails-issue-guide). To re-open this issue, please copy the template from [here](https://raw.githubusercontent.com/balderdashy/sails/master/.github/ISSUE_TEMPLATE), paste it at the beginning of your initial comment, and follow the instructions in the text. Then post a new comment (e.g. "ok, fixed!") so that I know to go back and check.

Sorry to be a hassle, but following these instructions ensures that we can help you in the best way possible and keep the Sails project running smoothly.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by MuslimMunir** [comment]
**Sails version**:0.12.4
**Node version**:6.11.4
**NPM version**:3.10.10
**DB adapter name**: sails-mysql
**DB adapter version**: 0.11.5
**Operating system**:ubuntu 14.04


<hr/>
I have observed that this error is also thrown when the sails app is halted for a while and the response is not being sent.
Like the case I have tried is I have given wrong mysql server credentials or the credentials of server to which i don't have access allowed.
What happened next is this error

Error occurred in session middleware :: 'Error: Handshake inactivity timeout\n at Handshake. (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:160:17)\n at emitNone (events.js:86:13)\n at Handshake.emit (events.js:185:7)\n at Handshake._onTimeout (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/sequences/Sequence.js:126:8)\n at tryOnTimeout (timers.js:232:11)\n at Timer.listOnTimeout (timers.js:202:5)\n --------------------\n at Protocol._enqueue (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:141:48)\n at Protocol.handshake (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:52:41)\n at PoolConnection.connect (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Connection.js:136:18)\n at Pool.getConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:48:16)\n at Pool.releaseConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:155:10)\n at Pool._removeConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:275:8)\n at Pool._purgeConnection (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:256:8)\n at Handshake.onConnect (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/Pool.js:57:14)\n at Handshake.Sequence.end (/home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/sequences/Sequence.js:85:24)\n at /home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:399:18\n at Array.forEach (native)\n at /home/muslimmunir/IdeaProjects/niche/sample/nicheacademy/node_modules/express-mysql-session/node_modules/mysql/lib/protocol/Protocol.js:398:13\n at _combinedTickCallback (internal/process/next_tick.js:67:7)\n at process._tickDomainCallback (internal/process/next_tick.js:122:9)'

and after that I tried to hit that same API with loadtest thousands of time and program crashed with below error.

_http_outgoing.js:344
throw new Error('Can't set headers after they are sent.');
^
Error: Can't set headers after they are sent.
at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:344:11)
at ServerResponse.res.setHeader (/usr/lib/node_modules/sails/node_modules//lib/patch.js:134:22)
at ServerResponse.res.set.res.header (/usr/lib/node_modules/sails/node_modules//lib/response.js:595:10)
at ServerResponse.res.send (/usr/lib/node_modules/sails/node_modules//lib/response.js:151:12)

**Comment by sailsbot** [comment]
Hi @MuslimMunir!  It looks like you may have removed some required elements from the initial comment template, without which I can't verify that this post meets our [contribution guidelines](http://bit.ly/sails-issue-guide). To re-open this issue, please copy the template from [here](https://raw.githubusercontent.com/balderdashy/sails/master/.github/ISSUE_TEMPLATE), paste it at the beginning of your initial comment, and follow the instructions in the text. Then post a new comment (e.g. "ok, fixed!") so that I know to go back and check.

Sorry to be a hassle, but following these instructions ensures that we can help you in the best way possible and keep the Sails project running smoothly.

*If you feel this message is in error, or you want to debate the merits of my existence (sniffle), please contact inquiries@sailsjs.com

**Comment by sgress454** [comment]
Duplicate of #4223
